langchain/src/document_loaders/web/notionapi.ts

```diff
@@ -114,55 +122,107 @@ export class NotionAPILoader extends BaseDocumentLoader {
     };
   }
 
-  private async loadPage(
-    page: string | PageObjectResponse
-  ): Promise<Document[]> {
-    // Check page is a page ID or a GetPageResponse
-    const [pageData, pageId] =
-      typeof page === "string"
-        ? [this.notionClient.pages.retrieve({ page_id: page }), page]
-        : [page, page.id];
+  private async loadBlock(block: BlockObjectResponse): Promise<MdBlock> {
+    return {
+      type: block.type,
+      blockId: block.id,
+      parent: await this.n2mClient.blockToMarkdown(block),
+      children: [],
+    };
+  }
 
-    const [pageDetails, pageBlocks] = await Promise.all([
-      pageData,
-      getBlockChildren(this.notionClient, pageId, null),
-    ]);
+  private async loadBlocks(
+    blocksResponse: ListBlockChildrenResponseResults
+  ): Promise<{ mdBlocks: MdBlock[]; childDocuments: Document[] }> {
+    const blocks = blocksResponse.filter(isFullBlock);
 
-    const [childPageDocuments, childDatabaseDocuments, mdBlocks] =
+    const [childPageDocuments, childDatabaseDocuments, blocksDocsArray] =
       await Promise.all([
         Promise.all(
-          pageBlocks
-            .filter(
-              (block) => "type" in block && block.type.includes("child_page")
-            )
+          blocks
+            .filter((block) => block.type.includes("child_page"))
             .map((block) => this.loadPage(block.id))
         ),
         Promise.all(
-          pageBlocks
+          blocks
+            .filter((block) => block.type.includes("child_database"))
+            .map((block) => this.loadDatabase(block.id))
+        ),
+        Promise.all(
+          blocks
             .filter(
-              (block) =>
-                "type" in block && block.type.includes("child_database")
+              (block) => !["child_page", "child_database"].includes(block.type)
+            )
+            .map((block) =>
+              (async () => {
+                const mdBlock = await this.loadBlock(block);
+                let childDocuments: Document[] = [];
+
+                if (block.has_children) {
```

Comment:
I feel like this logic should be recursive within `loadBlock`?

Reply:
Yeah, I agree.

The problem is that this will require a large refactor because of the way it loads `MdBlock[]` and `Document[]`.

I renamed `loadBlocks(blocks)` to `loadBlocksAndDocs(blocks)`. Hopefully that clears up any confusion for now.